timdex_dataset_api/record.py

```diff
@@ -19,35 +19,51 @@ class DatasetRecord:
     timdex_record_id: str
     source_record: bytes
     transformed_record: bytes
+    source: str
+    run_date: str | datetime.datetime
+    run_type: str
+    run_id: str
+    action: str
 
     # partition columns
-    source: str | None = None
-    run_date: str | datetime.datetime | None = None
-    run_type: str | None = None
-    run_id: str | None = None
-    action: str | None = None
+    year: str | None = None
+    month: str | None = None
+    day: str | None = None
```

Comment:
The partition columns are (a) optional since the `__post_init__` method will attempt to derive the partition values from the `run_date` and (b) are set as `str` objects to support zero-padded months and dates (e.g. `01`, `02`)

Reply:
Had a comment about agreeing that these are correctly optional, but I wonder if we could further ensure that `year`, `month`, and `day` are _only_ derived from `run_date`.

What if `year`, `month`, and `day` were all dynamic properties?  We would no longer requier a `__post_init__` method, as we'd derive this only when needed.

Then, when we serialize via `to_dict()`, we could ensure that these are included in the output and would therefore get used during writing.

This didn't seem obvious at first when this was discussed (or notes in engineering plan or tickets), but it's feeling like these are somewhat special.  I think we want to avoid at all costs a situation where they somehow differ from `run_date`, or `run_date` is present but these aren't (or vice versa).  By making them dynamic properties of `run_date`, it feels like we gaurd against all those situations.

We _could_ still exercise these properties via `__post_init__` if we wanted, to ensure that `run_date` is a) present, and b) provides meaningful partition values.  That may be worth considering.